File: Leetcode/typescript/1448-good-nodes.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function goodNodes(root: TreeNode | null): number {
    if (!root) {
        return 0;
    }
    function dfs(node: TreeNode | null, routeMax: number) {
        if (node === null) {
            return 0;
        }
        let currMax = Math.max(routeMax, node.val);
        let val = dfs(node.left, currMax) + dfs(node.right, currMax);
        if (node.val >= currMax) {
            val++;
        }
        return val;
    }
    return dfs(root, root.val);
}

File: Leetcode/python/1448-good-nodes.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def goodNodes(self, root: TreeNode) -> int:

        def dfs(node, routeMax):
            if not node:
                return 0
            currMax = max(routeMax, node.val)
            val = dfs(node.left, currMax) + dfs(node.right, currMax)
            if node.val >= currMax:
                val += 1
            return val

        return dfs(root, root.val)
